src/enzo/ActiveParticle_SpringelHernquist.C

```diff
@@ -0,0 +1,245 @@
+    if((*it)->name == "metallicity") (*it)->name = "metallicity_fraction";
+  }
+  
+  return SUCCESS;
+  
+}
+
+int ActiveParticleType_SpringelHernquist::EvaluateFormation
+(grid *thisgrid_orig, ActiveParticleFormationData &supp_data)
+{
+  SpringelHernquistGrid *tg =
+    static_cast<SpringelHernquistGrid *>(thisgrid_orig);
+
+  float bmass, tstar;
+  int i, j, k, index, offset_y, offset_z;
+  int NumberOfNewParticles = 0;
+  float r_float, y, x, pstar, starfraction, usn;
+  float msolar=SolarMass, mproton=mh, beta=0.1,
+    sqrtepssn=2.0e24, kb=kboltz;
+
+  // Define the energy output from supernovae
+
+  usn = (1. - beta) * sqrtepssn / beta / msolar; // Below Eq (2)
+  usn = usn * sqrtepssn; // this trick done to avoid floating-point issues (BWO)
+
+  /* Make it pretty */
+
+  float *density = tg->BaryonField[supp_data.DensNum];
+  float *velx = tg->BaryonField[supp_data.Vel1Num];
+  float *vely = tg->BaryonField[supp_data.Vel2Num];
+  float *velz = tg->BaryonField[supp_data.Vel3Num];
+
+  bool HasMetalField = (supp_data.MetalNum != -1 ||
+			supp_data.ColourNum != -1);
+
+  int GridDimension[3] = {tg->GridDimension[0],
+                          tg->GridDimension[1],
+                          tg->GridDimension[2]};
+
+  // Pre-calculate serialized offsets for the 3D data field.  Used for
+  // the divergence.
+  offset_y = tg->GridDimension[0];
+  offset_z = tg->GridDimension[0] * tg->GridDimension[1];
+
+  for (k = tg->GridStartIndex[2]; k <= tg->GridEndIndex[2]; k++) {
+    for (j = tg->GridStartIndex[1]; j <= tg->GridEndIndex[1]; j++) {
+      index = GRIDINDEX_NOGHOST(tg->GridStartIndex[0], j, k);
+      for (i = tg->GridStartIndex[0]; i <= tg->GridEndIndex[0]; i++, index++) {
+
+	// 0. If no more room for particles, quit.
+	if (supp_data.NumberOfNewParticles >=
+	    supp_data.MaxNumberOfNewParticles)
+          continue;
+
+	// 1. Finest level of refinement
+	if (tg->BaryonField[tg->NumberOfBaryonFields][index] != 0.0) 
+	  continue;
+	
+	// 2. Density greater than threshold
+	if (density[index] < OverDensityThreshold)
+	  continue;
+
+	// Calculate star formation timescale. Eq 21.
+	tstar =  31556926.0 * MinimumDynamicalTime * 
+	  pow(density[index] * supp_data.DensityUnits / PhysicalDensityThreshold, -0.5);
+
+	/* note: minus sign is because 'coolrate' is actual backwards: when coolrate is 
+	 * negative, gas is cooling (which is the opposite of how it's defined in Springel
+	 * & Hernquist 2003, MNRAS, 339, 289, eqtn. 16 */
+	y = -1.0 * tstar * supp_data.CoolingRate[index] /
+	  (density[index] * supp_data.DensityUnits) /
+	  (beta * usn - (1.0 - beta) * 1.5 *
+	   kb * supp_data.Temperature[index] / 0.6 / mproton);
+
+	// Calculate the fraction of mass in cold clouds.
+	if (y <= 0) {
+	  x = 0;
+	} else {
+	  x = 1. + 1./ 2. / y - sqrt(1. / y + 1. / 4. / y / y); // Eq(18)
+	}
+
+	// Calculate total baryon mass in the cell.
+	bmass = density[index] * supp_data.MassUnits;
```

Comment:
baryon mass = density*mass unit??????

Reply:
Non-intuitive that's for sure. But MassUnits is defined in ActiveParticle.C as 
`data.DensityUnits*POW(data.LengthUnits,3);`